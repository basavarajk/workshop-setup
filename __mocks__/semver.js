module.exports = {
  satisfies: jest.fn(),
}
