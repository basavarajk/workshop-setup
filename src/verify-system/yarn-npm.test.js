/* eslint no-console:0 */
import cpMock from 'child_process'
import {oneLine} from 'common-tags'
import {satisfies as satisfiesMock} from 'semver'
import validateYarnNpm from './yarn-npm'

jest.mock('child_process')

test('returns null if conditions are satisfied', () => {
  setup({satisfiesYarn: true})
  const desiredYarn = '^0.21.3'
  const result = validateYarnNpm(desiredYarn)()
  expect(result).toBe(null)
})

test('returns yarn mismatch error when yarn version is wrong', () => {
  const mockYarnVersion = '0.21.3'
  setup({satisfiesYarn: false, mockYarnVersion})
  const desiredYarn = '^0.20.3'
  const result = validateYarnNpm(desiredYarn)()
  expect(result).toContain(desiredYarn)
  expect(result).toContain(mockYarnVersion)
  expect(result).toMatchSnapshot()
})

test('returns yarn is not installed error when yarn version is wrong', () => {
  setup({
    mockYarnVersion: () => {
      throw new Error('blah')
    },
  })
  const desiredYarn = '^0.20.3'
  const desiredNpm = '^4.0.3'
  const result = validateYarnNpm(desiredYarn, desiredNpm)()
  expect(result).toContain('install yarn')
  expect(result).toMatchSnapshot()
})

test('returns npm version mismatch', () => {
  const mockNpmVersion = '3.11.3'
  setup({
    mockNpmVersion,
    satisfiesNpm: false,
    mockYarnVersion: () => {
      throw new Error('blah')
    },
  })
  const desiredYarn = '^0.20.3'
  const desiredNpm = '^4.0.3'
  const result = validateYarnNpm(desiredYarn, desiredNpm)()
  expect(result).toContain(desiredNpm)
  expect(result).toContain(mockNpmVersion)
  expect(result).toContain('use npm')
  expect(result).toMatchSnapshot()
})

function setup(
  {
    mockYarnVersion = '0.21.3',
    mockNpmVersion = '4.0.3',
    satisfiesYarn = true,
    satisfiesNpm = true,
  } = {},
) {
  satisfiesMock.mockClear()
  cpMock.execSync.mockClear()
  satisfiesMock.mockImplementation(satisfiesMockImplementation)
  cpMock.execSync.mockImplementation(execSyncMock)

  function satisfiesMockImplementation(actual, desired) {
    if (actual === mockYarnVersion) {
      return satisfiesYarn
    } else if (actual === mockNpmVersion) {
      return satisfiesNpm
    } else {
      console.error(
        oneLine`
          called satisifies with "${desired}, ${actual}".
          ${actual} does not match yarn@${mockYarnVersion}
          or npm@${mockNpmVersion}
        `,
      )
      throw new Error('satisfies mock implementation insufficient')
    }
  }

  function execSyncMock(command) {
    if (command.includes('yarn')) {
      return callAndReturn(mockYarnVersion)
    } else if (command.includes('npm')) {
      return callAndReturn(mockNpmVersion)
    } else {
      console.error(
        oneLine`
          called execSync with "${command}".
          This does not include yarn or npm
        `,
      )
      throw new Error('execSync mock implementation insufficient')
    }
  }

  function callAndReturn(thing) {
    if (typeof thing === 'function') {
      thing = thing()
    }
    return thing
  }
}
