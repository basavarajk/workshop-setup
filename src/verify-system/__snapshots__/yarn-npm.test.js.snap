// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns npm version mismatch 1`] = `
"You do not have yarn installed. This is a package manager client that installs from the regular npm registry, but ensures you get the same versions of all dependencies required for this repository. It is highly recommended that you install yarn: \`npm install --global yarn@^0.20.3\` (learn more: https://yarnpkg.com/)
Note that you can also ignore this warning and continue with npm if you prefer.
Your version of npm (3.11.3) does not satisfy the desired range of ^4.0.3. You should install yarn anyway, but if you would rather use npm, please at least have a version within the specified version range. You can install the latest version by running \`npm install --global npm@^4.0.3\`."
`;

exports[`returns yarn is not installed error when yarn version is wrong 1`] = `
"You do not have yarn installed. This is a package manager client that installs from the regular npm registry, but ensures you get the same versions of all dependencies required for this repository. It is highly recommended that you install yarn: \`npm install --global yarn@^0.20.3\` (learn more: https://yarnpkg.com/)
Note that you can also ignore this warning and continue with npm if you prefer."
`;

exports[`returns yarn mismatch error when yarn version is wrong 1`] = `"Your version of yarn (0.21.3) does not satisfy the desired range of ^0.20.3. Run \`yarn self-update\` (or \`npm install --global yarn@^0.20.3\`) to update."`;
